### DISCUSSION TITLE
Phase Field GrainGrowth simulation coupled with temperature Gradient

### DISCUSSION THREAD

**Discussion by zengfy-hust** [idaholab/moose#18260] [Q&A Modules: Phase field] [answered]
Hello MOOSE：
         I　recently　make some effort in couping grainGrowth with temperature ,however ,some problems occured when I set the temperature gradient like following:
  [./TGradient]
    type = ParsedFunction
    value = '450 + 0.12*x'
  [../] 
the following error occured after one perfect time step :
![2](https://user-images.githubusercontent.com/56660568/124477192-51337a00-ddd6-11eb-9ab8-28e1c8d71a78.png)
![1](https://user-images.githubusercontent.com/56660568/124477187-4f69b680-ddd6-11eb-895c-e2bf249e7119.png)
when I changed the value to 450 + 0.02*x,the simulation workes fine .could you tell me what is the peoblems ? 
any useful suggestions will be fine .

**Comment by GiudGiud** [comment, marked as answer]
Hello

Sorry we seemed to have missed this post. 
@jiangwen84  or @laagesen anything jumping to mind on this?

Guillaume

**Reply by SudiptaBiswas** [reply]
The error message is coming from `GrainTracker`. Probably something unphysical is happening at the higher value.

**Reply by zengfy-hust** [reply]
For your better understanding of my question ,I provide the input file I used here .Hope you could provide some help  
@GiudGiud @SudiptaBiswas 
[1-350-70-texture.txt](https://github.com/idaholab/moose/files/6852526/1-350-70-texture.txt)
[grain_growth_temp_gradient.txt](https://github.com/idaholab/moose/files/6852536/grain_growth_temp_gradient.txt)

**Reply by GiudGiud** [reply]
I cant help, I dont know much about grain growth. @SudiptaBiswas anyone else we could tag here
@zengfy-hust anything unphysical in the model outputs (print material properties, aux variables etc in an exodus file) before the failure?

**Reply by SudiptaBiswas** [reply]
@zengfy-hust What does your grain structure and mesh look like? Clearly, you have a size issue in there. Why do you have split gains? They definitely could be the source of the error. Could you reproduce the error on a smaller/simpler problem? Couple of things you should try, 
1. Run the thermal gradient simulations without the EBSD reader. Try using Voronoi structure as a starting point. This would ensure that nothing unphysical is happening at that temperature. 
2. Use finer mesh and proper threshold for GrainTracker.  Try to avoid split grains.

**Reply by zengfy-hust** [reply]
thank you for your help ,I will work on the two options now ,and see if either of it  works.

**Reply by dschwen** [reply]
@GiudGiud how do you decide who to tag in discussions? Do you look at the git blame to find out who wrote the code in question?

**Reply by GiudGiud** [reply]
Not always, I try to vary the people. Sometimes I get it wrong and tag people on a system they're not familiar with. My hope is they tag the right people after

**Comment by dschwen** [comment]
It looks like what's happening here is the subdivision of a grain into two grains. That new grain will have the next available ID, which is 281 in your case. The EBSDReader has no information (like Euler angles) on that new grain however, because it is not in the initial datafile. 

Without seeing your input I must assume that you are using the EBSDReader as an `euler_angle_provider`. You need to create a new object derived from `EulerAngleProvider` and `GrainDataTracker<T>` that will work as your new grain tracker and Euler angle provider. In that class you can override the `newGrain` method to assign a new set of Euler angles to the new grain. You will also need to override `getEulerAngles` to return the angles from your EBSDReader (needs to be a parameter of this object) if the grain ID is below the `getGrainNum()` of the EBSDReader, and return your randomly created Euler angles otherwise.

The `newGrain` method you implement will automatically get called by the grain tracking system if a new previously unknown grain is discovered.

Note that for now the new grain data cannot be connected to a previous grain, because currently we have no way of determining the pre-split grain ID. Maybe that is something we (or you) could work on (i.e to assign the split-off part the same orientation as the original grain).
